theseus/embodied/robot/forward_kinematics.py

```diff
@@ -0,0 +1,146 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+from typing import List, Optional
+
+from theseus.labs.lie_functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+
+# TODO: Add support for joints with DOF>1
+
+
+def ForwardKinematicsFactory(robot: Robot, link_names: Optional[List[str]] = None):
+    links: List[Link] = []
+
+    if link_names is None:
+        links = robot.links
+    else:
+        links = [robot.link_map[name] for name in link_names]
+
+    link_ids: List[int] = [link.id for link in links]
+
+    ancestors: List[Link] = []
+    joint_ids: List[int] = []
+    for link in links:
+        ancestors += [anc for anc in link.ancestors]
+        joint_ids += link.angle_ids
+    pose_ids = sorted(list(set([anc.id for anc in ancestors] + link_ids)))
+    joint_ids = sorted(list(set(joint_ids)))
+
+    def _forward_kinematics_helper(angles: torch.Tensor):
+        if angles.ndim != 2 or angles.shape[1] != robot.dof:
+            raise ValueError(
+                f"Joint angles for {robot.name} should be {robot.dof}-D vectors"
```

Comment:
Maybe add "{robot.dof}-D vectors with a batch dimension."?

Reply:
Not sure. If I remember correctly, we only consider the second dimension in our Lie group code.

Reply:
I'm just suggesting a more descriptive error message for what your code is already doing (you are checking ndim == 2 and dof being in dim 1). Your current message implies that `ndim` should be `==1`.